torch_xla/csrc/tensor_methods.cpp

```diff
@@ -957,25 +969,54 @@ XLATensor XLATensor::diagonal(const XLATensor& input, xla::int64 offset,
 
 XLATensor XLATensor::div(const XLATensor& input, const XLATensor& other,
                          c10::optional<at::ScalarType> logical_element_type) {
-  return input.CreateFrom(input.GetIrValue() / other.GetIrValue(),
-                          logical_element_type);
+  at::ScalarType scalar_type =
+      at::typeMetaToScalarType(c10::get_default_dtype());
+  xla::PrimitiveType input_type = input.GetIrValue().shape().element_type();
+  xla::PrimitiveType other_type = other.GetIrValue().shape().element_type();
+  bool input_is_float = xla::primitive_util::IsFloatingPointType(input_type);
+  bool other_is_float = xla::primitive_util::IsFloatingPointType(other_type);
+  if (input_is_float && !other_is_float) {
+    scalar_type = TensorTypeFromXlaType(input_type);
+  } else if (!input_is_float && other_is_float) {
+    scalar_type = TensorTypeFromXlaType(other_type);
+  }
+  ir::Value input_value = GetFloatingIrValue(input, scalar_type);
```

Comment:
Make the `GetFloatingIrValue()` API to return an `std::pair<ir::Value, ir::Value>`, returning the original and possibly casted values.
Then get the original type from the original value shape.

Reply:
I think changing the `GetFloatingIrValue()` api will cause additional expense to the `div(const XLATensor& input, at::Scalar other` and `div_` as they don't need to know the original type. 

What's the advantage of updating the api to return the pair? I think we need to know the `scalar_type` before calling `GetFloatingIrValue` and we don't really care about the original type after that right?

Reply:
Hmm, either the code changed or I mis-read it 😄 
No need to change that API.